jupyterlab_gcedetails/src/components/hardware_scaling_form.tsx

```diff
@@ -0,0 +1,243 @@
+/**
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as csstips from 'csstips';
+import * as React from 'react';
+
+import {
+  css,
+  SelectInput,
+  CheckboxInput,
+  LearnMoreLink,
+  BASE_FONT,
+  ActionBar,
+  SubmitButton,
+} from 'gcp_jupyterlab_shared';
+import { stylesheet, classes } from 'typestyle';
+import { SynchronizedSliders } from './synchronized_sliders';
+import {
+  ACCELERATOR_COUNTS_1_2_4_8,
+  machineTypes,
+  ACCELERATOR_TYPES,
+} from '../data';
+
+interface Props {
+  // Assumes sorted list with memory and cpu set to 0 for first value
+  onDialogClose?: () => void;
+}
+
+interface State {
+  baseMachineType: string;
+  cpuValue: number;
+  memoryValue: number;
+  attachGpu: boolean;
+  gpuType?: string;
+  gpuCount?: string;
+}
+
+export const STYLES = stylesheet({
+  checkbox: {
+    marginRight: '10px',
+  },
+  checkboxContainer: {
+    paddingTop: '8px',
+    paddingBottom: '10px',
+  },
+  title: {
+    ...BASE_FONT,
+    fontWeight: 500,
+    fontSize: '15px',
+    marginBottom: '5px',
+    ...csstips.horizontal,
+    ...csstips.flex,
+  },
+  formContainer: {
+    padding: '26px 16px 0px 16px',
+  },
+  container: {
+    width: '500px',
+  },
+  description: {
+    paddingBottom: '20px',
+  },
+  topPadding: {
+    paddingTop: '10px',
+  },
+  bottomPadding: {
+    paddingBottom: '10px',
+  },
+});
+
+const NO_ACCELERATOR = ACCELERATOR_TYPES[0].value as string;
+
+export class HardwareScalingForm extends React.Component<Props, State> {
+  constructor(props: Props) {
+    super(props);
+
+    this.state = {
+      baseMachineType: machineTypes[0].base.value as string,
+      cpuValue: 0,
+      memoryValue: 0,
+      attachGpu: false,
+      gpuType: NO_ACCELERATOR,
+      gpuCount: '',
+    };
+  }
+
+  private getMachineType = (baseMachineType: string) => {
+    return machineTypes.find(elem => elem.base.value === baseMachineType);
+  };
+
+  render() {
+    const { onDialogClose } = this.props;
+    const { gpuType, gpuCount, attachGpu, baseMachineType } = this.state;
+
+    const onBaseMachineTypeChange = (
```

Comment:
I'd suggest lifting these out to private methods of the class rather than lambdas defined during each call of the `render` method. You may even be able to generalize a method that works for the Cpu, Memory, and GpuCount fields which all take a single value and assign a state property.

Reply:
That's a good point I'll remove the lambdas.
For the generalized method, do you mean something like this:
`  private onChange(key: string, newValue: number) {
    this.setState({
      [String(key)]: newValue,
    } as unknown);
  }`
Or could it also make sense to call setState directly in the onChange props for each of the components in the render method, like [this](https://github.com/GoogleCloudPlatform/jupyter-extensions/blob/c1fb81d9c22575255dd372991aeda5df0321aed0/jupyterlab_gcpscheduler/src/components/schedule_builder/hour_schedule_form.tsx#L90)? Not sure of best practice here :/

Reply:
I think either option is reasonable. I think the latter option looks a bit more TS-friendly compared to having to cast the type as `unknown`.